File: src/consume-block-comment.js
/**
 * @fileoverview
 * Consumes a block comment from JSONC input.
 */

/**
 * @typedef {import('./tokenize-jsonc.js').Token} Token
 * @typedef {import('./tokenize-jsonc.js').TokenizeError} TokenizeError
 */

/**
 * Result from consumeBlockComment.
 * @typedef {Object} ConsumeResult
 * @property {number} column - Updated column position
 * @property {number} line - Updated line number
 * @property {number} pos - Updated position in the string
 * @property {Token} token - The consumed comment token
 */

/**
 * Consumes a block comment.
 * @param {string} jsonc - The input string
 * @param {number} startPos - Starting position (at the first '/')
 * @param {number} len - Length of input
 * @param {number} startLine - Starting line number
 * @param {number} startColumn - Starting column
 * @param {TokenizeError[]} errors - Errors array to append to
 * @returns {ConsumeResult}
 */
export function consumeBlockComment(jsonc, startPos, len, startLine, startColumn, errors) {
    let pos = startPos + 2; // skip '/*'
    let line = startLine;
    let column = startColumn + 2;
    let value = '/*';

    // Consume until '*/' or end of input.
    while (pos < len) {
        const ch = jsonc[pos];

        // Check for closing '*/'.
        if (ch === '*' && pos + 1 < len && jsonc[pos + 1] === '/') {
            value += '*/';
            pos += 2;
            column += 2;
            return {
                column,
                line,
                pos,
                token: { category: 'comment', column: startColumn, line: startLine, value },
            };
        }

        // Handle newlines inside block comment.
        if (ch === '\n') {
            value += ch;
            pos += 1;
            line += 1;
            column = 0;
            continue;
        }
        if (ch === '\r') {
            value += ch;
            pos += 1;
            // Handle \r\n.
            if (pos < len && jsonc[pos] === '\n') {
                value += '\n';
                pos += 1;
            }
            line += 1;
            column = 0;
            continue;
        }

        // Regular character.
        value += ch;
        pos += 1;
        column += 1;
    }

    // Reached end without closing '*/'.
    errors.push({
        category: 'syntax',
        column: startColumn,
        line: startLine,
        message: 'Unterminated block comment',
    });

    return {
        column,
        line,
        pos,
        token: { category: 'comment', column: startColumn, line: startLine, value },
    };
}

File: wasm/jsonc_lexer/src/consume_block_comment.rs
//! Consumes a block comment from JSONC input.

use crate::types::{ConsumeResult, Token, TokenizeError};

/// Consumes a block comment (/* ... */).
pub fn consume_block_comment(
    bytes: &[u8],
    start_pos: usize,
    len: usize,
    start_line: u32,
    start_column: u32,
    errors: &mut Vec<TokenizeError>,
) -> ConsumeResult {
    let mut pos = start_pos + 2; // skip '/*'
    let mut line = start_line;
    let mut column = start_column + 2;
    let mut value = vec![b'/', b'*'];

    // Consume until '*/' or end of input.
    while pos < len {
        let ch = bytes[pos];

        // Check for closing '*/'.
        if ch == b'*' && pos + 1 < len && bytes[pos + 1] == b'/' {
            value.push(b'*');
            value.push(b'/');
            pos += 2;
            column += 2;
            return ConsumeResult {
                column,
                line,
                pos,
                token: Token {
                    category: "comment".to_string(),
                    column: start_column,
                    line: start_line,
                    value: String::from_utf8_lossy(&value).to_string(),
                },
            };
        }

        // Handle newlines inside block comment.
        if ch == b'\n' {
            value.push(ch);
            pos += 1;
            line += 1;
            column = 0;
            continue;
        }
        if ch == b'\r' {
            value.push(ch);
            pos += 1;
            // Handle \r\n.
            if pos < len && bytes[pos] == b'\n' {
                value.push(b'\n');
                pos += 1;
            }
            line += 1;
            column = 0;
            continue;
        }

        // Regular character.
        value.push(ch);
        pos += 1;
        column += 1;
    }

    // Reached end without closing '*/'.
    errors.push(TokenizeError {
        category: "syntax".to_string(),
        column: start_column,
        line: start_line,
        message: "Unterminated block comment".to_string(),
    });

    ConsumeResult {
        column,
        line,
        pos,
        token: Token {
            category: "comment".to_string(),
            column: start_column,
            line: start_line,
            value: String::from_utf8_lossy(&value).to_string(),
        },
    }
}
